### ISSUE TITLE
[Clang] Friend function reported to be undeclared unless one of the arguments is the class type

### ISSUE THREAD

**Issue by ShawnZhong** [llvm/llvm-project#60524]
In the example below, `foo` is reported to be undeclared unless one of the arguments is the class type. 

```cpp
struct S {
  friend void foo(){};
  friend void foo(void*){};
  friend void foo(void*, void*){};
};

void test() {
  foo();  // use of undeclared identifier 'foo'

  foo(nullptr);      // use of undeclared identifier 'foo'
  foo((S*)nullptr);  // no error

  foo(nullptr, nullptr);          // use of undeclared identifier 'foo'
  foo((S*)nullptr, nullptr);      // no error
  foo(nullptr, (S*)nullptr);      // no error
  foo((S*)nullptr, (S*)nullptr);  // no error
}
```

I think that in all the cases, `foo` should be declared (and also defined), so the program should be valid. 
https://eel.is/c++draft/class.friend#example-6 provides an example of defining friend function within the class.

**Comment by shafik** [comment]
I think if I read [class.friend p6](https://eel.is/c++draft/class.friend#example-6) and [class.friend p11](https://eel.is/c++draft/class.friend#11) it sure seems that way.

but I think the key here is [decl.meaning.general p2.1](https://eel.is/c++draft/dcl.meaning#general-2.1) which says:

> If the declaration is a friend declaration:
> - (2.1) The declarator does not bind a name.

clang/gcc/MSVS all seem to agree here: https://godbolt.org/z/fT9fse7K5

I do feel like the wording could use some clarification and an example that demonstrates this case seems warranted.

CC @hubert-reinterpretcast for conformance check-in
